### ISSUE TITLE
buck not reusing cache across flavours

### ISSUE THREAD

**Issue by nikhedonia** [facebook/buck#2052]
I've identified a missed caching opportunity:

Given the following .buckconfig:  

```
[cxx#foo]
[cxx#bar]
```

`buck build :main#foo` won't use the build and cached artefacts of  `buck build :main#bar` and vice-versa.

This is very unfortunate as the flavours might have a lot incommon.

**Comment by nikhedonia** [comment]
> Flavors usually change the behaviour, so, unless they are not being treated in a special way (e.g. stripped inside corresponding Description), they affect the output path at minimum

We don't know what the different flavors will do but if we would only cache based on hashes of all involved files, options and env variables(excluding flavor name) we would have less cache misses.

An example - quite common in c++ - are optional features, where we swap one implementation with another eg.:

 ```
cxx_library(
  name = "foo",
  deps = [":a, ":b"] if myCustomOption else ['":a"] 
  srcs = srcs
);
```

I also think it is interesting to note if you change options within a flavor, the caching works correctly eg.:

```
[cxx#foo]
  myCustomOption=  1 
[cxx#bar]
  myCustomOption= 0 
```

changing `myCustomOption=0`  will correctly reuse the cache where possible.

**Comment by beefon** [comment]
> excluding destination / flavor where possible

That is the key, this requires a knowledge about any potential flavor. Buck treats some flavors internally in a special way and skips them if needed. These flavors are actually internal ones in most cases.

> if we would only cache based on hashes of all involved files, options(excluding destination / flavor where possible) and env variables we would have less cache misses

Yes, but the chance of getting a corrupted/incorrectly built binaries will also increase.

**Comment by beefon** [comment]
I think the better approach in your case could be to introduce two different targets with different deps, so then you can build them like `buck build :main_foo` and `buck build :main_bar`. In this case, if both targets depend on `A`, its cached output should be correctly reused.

**Comment by nikhedonia** [comment]
> I think the better approach in your case could be to introduce two different targets with different deps, so then you can build them like buck build :main_foo and buck build :main_bar

Not really, in some libraries we have more than 10 binary options. This approach would result in more than >2^10 different variations foreach rule. This is not feasible nor user friendly.

**Comment by beefon** [comment]
> Can an user defined flavor change the action graph?

It depends, different codepaths either propagate the passed flavors down to the deps or not, there is no consistency. I think this is still the case, but Buck maintainers may have changed this already. You may notice there is no documentation about the flavors, so there should not be any expectation or assumption how do they work internally. This is unfortunate, but this is reality.

I totally agree with you: this could be implemented in a better and more understandable and predictable way. I guess if you have any imagination what can be done in order to make flavors better or even remove them, please share your thoughts. When I was on a Buck team, we raised this a couple of times and there was not clear way of fixing this. It was all messy, and I think it is still like that, but again, something may have been improved over the last year or so. :)

> In some libraries we have more than 10 binary options. This approach would result in more than >2^10 different variations foreach rule.

Then this is definitely not a solution for you.
